presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,250 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.server;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.prestosql.execution.SqlQueryExecution;
+import io.prestosql.execution.StageInfo;
+import io.prestosql.execution.StageState;
+import io.prestosql.execution.TaskInfo;
+import io.prestosql.execution.TaskManagerConfig;
+import io.prestosql.spi.QueryId;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.DynamicFilters;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.SymbolReference;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.Immutable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+@ThreadSafe
+public class DynamicFilterService
+{
+    @GuardedBy("this")
+    private final Map<SourceDescriptor, SettableFuture<Domain>> dynamicFilterSummaries = new HashMap<>();
+
+    @GuardedBy("this")
+    private final Map<QueryId, Supplier<List<StageInfo>>> queries = new HashMap<>();
+
+    private final ScheduledExecutorService collectDynamicFiltersExecutor = newSingleThreadScheduledExecutor(daemonThreadsNamed("DynamicFilterService"));
+
+    @Inject
+    public DynamicFilterService(TaskManagerConfig taskConfig)
+    {
+        collectDynamicFiltersExecutor.scheduleWithFixedDelay(this::collectDynamicFilters, 0, taskConfig.getStatusRefreshMaxWait().toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
Instead of polling the tasks, the dynamic filters should be collected as they show up in the tasks (i.e., when HttpRemoteTask updates the state of the task)

Reply:
> Instead of polling the tasks, the dynamic filters should be collected as they show up in the tasks (i.e., when HttpRemoteTask updates the state of the task)

Making it poll based was simpler to implement initially (dynamic filter service state is much simpler) and is actually not much worse than event-based (since fetching task status is also poll based). Later we could implement event-based notification, but given https://github.com/prestosql/presto/pull/1072#discussion_r408356989 we might need to thing about it more.